tests/Unit/NumericalAlgorithms/LinearSolver/Test_DistributedConjugateGradientAlgorithm.cpp

```diff
@@ -0,0 +1,290 @@
+        Ap_global.begin() + (array_index + 1) * int(number_of_grid_points),
+        Ap_local.begin());
+    db::mutate<db::add_tag_prefix<
+        LinearSolver::Tags::OperatorAppliedTo,
+        db::add_tag_prefix<LinearSolver::Tags::Operand, ScalarFieldTag>>>(
+        make_not_null(&box), [&Ap_local](auto Ap) noexcept {
+          *Ap = Scalar<DataVector>(Ap_local);
+        });
+    // Proceed with algorithm
+    Parallel::get_parallel_component<ParallelComponent>(cache)[array_index]
+        .set_terminate(false);
+    Parallel::get_parallel_component<ParallelComponent>(cache)[array_index]
+        .perform_algorithm();
+  }
+};
+
+// Checks for the correct solution after the algorithm has terminated.
+struct TestResult {
+  template <
+      typename... DbTags, typename... InboxTags, typename Metavariables,
+      typename ActionList, typename ParallelComponent,
+      Requires<tmpl2::flat_any_v<cpp17::is_same_v<VariablesTag, DbTags>...>> =
+          nullptr>
+  static auto apply(db::DataBox<tmpl::list<DbTags...>>& box,
+                    tuples::TaggedTuple<InboxTags...>& /*inboxes*/,
+                    Parallel::ConstGlobalCache<Metavariables>& /*cache*/,
+                    const int array_index, const ActionList /*meta*/,
+                    const ParallelComponent* const /*meta*/) noexcept {
+    const auto& result = get<ScalarFieldTag>(box).get();
+    for (size_t i = 0; i < number_of_grid_points; i++) {
+      SPECTRE_PARALLEL_REQUIRE(
+          result[i] == approx(gsl::at(expected_results, array_index)[i]));
+    }
+  }
+};
+
+using linear_solver = LinearSolver::ConjugateGradient<VariablesTag>;
+
+struct InitializeElement {
+  using return_tag_list =
+      tmpl::append<tmpl::list<VariablesTag>,
+                   typename linear_solver::tags::simple_tags,
+                   typename linear_solver::tags::compute_tags>;
+
+  template <typename... InboxTags, typename Metavariables, typename ActionList,
+            typename ParallelComponent>
+  static auto apply(
+      const db::DataBox<tmpl::list<>>& /*box*/,
+      const tuples::TaggedTuple<InboxTags...>& /*inboxes*/,
+      const Parallel::ConstGlobalCache<Metavariables>& cache,
+      const int array_index, const ActionList /*meta*/,
+      const ParallelComponent* const parallel_component_meta) noexcept {
+    auto box = db::create<db::AddSimpleTags<tmpl::list<VariablesTag>>>(
+        VariablesType{number_of_grid_points, 0.});
+
+    auto linear_solver_box = linear_solver::tags::initialize(
+        std::move(box), cache, array_index, parallel_component_meta,
+        gsl::at(sources, array_index),
+        db::item_type<db::add_tag_prefix<LinearSolver::Tags::OperatorAppliedTo,
+                                         VariablesTag>>{number_of_grid_points,
+                                                        0.});
+
+    return std::make_tuple(std::move(linear_solver_box));
+  }
+};  // namespace
+
+template <typename Metavariables>
```

Comment:
It looks like the code from here down could be factored into an `hpp` and shared between the two tests. (might need to add some template parameters or magic values)

Reply:
Yes I intend to do that in the GMRES linear solver PR (coming up next, just need to refactor it in response to the changes in this PR). It can also reuse a lot of this code. Then I'll know better how to refactor.